### PR TITLE
8319813: Remove upper limit on number of compiler phases in phasetype.hpp

### PR DESCRIPTION
**Pull request by dlunde**
This changeset removes the implicit upper limit on the number of compiler phases in `phasetype.hpp`. The limit was due to the 64-bit mask used in `Compile::should_print_phase`, causing the `assert(cpt < 64, "out of bounds");` in `phasetype.hpp` to trigger when increasing the number of phases to more than 64.

Changes:
- Replace the 64-bit mask with a bit map (`utilities/bitMap.hpp`).
- Clean up the `PhaseNameValidator` interface by allocating the mask internally.
- Move the check for whether a mask is non-zero ("is set") into `PhaseNameValidator`.
- Add a method `should_print_phase` to `DirectiveSet`, simplifying the `if`-condition in `Compile::should_print_phase`.

### Testing
Platforms: windows-x64, linux-x64, linux-aarch64, macosx-x64, macosx-aarch64.
- `tier1`
- HotSpot parts of `tier2` and `tier3`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319813](https://bugs.openjdk.org/browse/JDK-8319813): Remove upper limit on number of compiler phases in phasetype.hpp (**Enhancement** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16729/head:pull/16729` \
`$ git checkout pull/16729`

Update a local copy of the PR: \
`$ git checkout pull/16729` \
`$ git pull https://git.openjdk.org/jdk.git pull/16729/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16729`

View PR using the GUI difftool: \
`$ git pr show -t 16729`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16729.diff">https://git.openjdk.org/jdk/pull/16729.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16729#issuecomment-1818509072)

### CODE REVIEW

File: src/hotspot/share/opto/phasetype.hpp [hunk 1]
```diff
@@ -157,11 +156,19 @@ class PhaseNameIter {
 
 class PhaseNameValidator {
  private:
+  CHeapBitMap _mask;
   bool _valid;
+  bool _set;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phasetype.hpp:161]: I suggest to also name this field `_mask_set` to match the accessor function `is_mask_set()`. 
```suggestion
  bool _mask_set;
```

File: src/hotspot/share/opto/phasetype.hpp [hunk 2]
```diff
@@ -172,10 +179,12 @@ class PhaseNameValidator {
         strncpy(_bad, *iter, len);
         _valid = false;
       } else if (PHASE_ALL == cpt) {
-        mask = ~(UINT64_C(0));
+        _mask.set_range(0,PHASE_NUM_TYPES);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phasetype.hpp:182]:
```suggestion
        _mask.set_range(0, PHASE_NUM_TYPES);
```

**Review by chhagedorn — APPROVED** [review]
Looks good!

File: src/hotspot/share/opto/phasetype.hpp [hunk 3]
```diff
@@ -186,13 +195,10 @@ class PhaseNameValidator {
...
-  bool is_valid() const {
-    return _valid;
-  }
-
-  const char* what() const {
-    return _bad;
-  }
+  const BitMap& mask() const { return _mask; }
+  bool is_valid() const { return _valid; }
+  bool is_mask_set() const { return _set; }
```
Comment (eme64) [inline src/hotspot/share/opto/phasetype.hpp:200]: Why not use `_mask.count_one_bits()`? Then you can drop the `_set` field.

File: src/hotspot/share/compiler/directivesParser.cpp [hunk 4]
```diff
@@ -336,12 +336,11 @@ bool DirectivesParser::set_option_flag(JSON_TYPE t, JSON_VAL* v, const key* opti
...
           }
         } else if (strncmp(option_key->name, "PrintIdealPhase", 15) == 0) {
-          uint64_t mask = 0;
-          PhaseNameValidator validator(s, mask);
+          PhaseNameValidator validator(s);
 
           valid = validator.is_valid();
           if (valid) {
-            set->set_ideal_phase_mask(mask);
+            set->set_ideal_phase_mask(validator);
```
Comment (eme64) [inline src/hotspot/share/compiler/directivesParser.cpp:343]: Why not pass it the mask directly?

File: src/hotspot/share/opto/phasetype.hpp [hunk 5]
```diff
@@ -157,11 +156,19 @@ class PhaseNameIter {
...
   char* _bad;
 
+
  public:
-  PhaseNameValidator(ccstrlist option, uint64_t& mask) : _valid(true), _bad(nullptr) {
+  PhaseNameValidator(ccstrlist option)
+                    : _mask(PHASE_NUM_TYPES, mtCompiler),
+                      _valid(true),
+                      _set(false),
+                      _bad(nullptr)
```
Comment (eme64) [inline src/hotspot/share/opto/phasetype.hpp:170]: I think the colon belongs up on the same line as `PhaseNameValidator(ccstrlist option)`, and the initializations only should be indented by 2 whitespaces.

File: src/hotspot/share/opto/phasetype.hpp [hunk 6]
```diff
@@ -186,13 +195,10 @@ class PhaseNameValidator {
...
   }
 
-  bool is_valid() const {
-    return _valid;
-  }
-
-  const char* what() const {
-    return _bad;
-  }
+  const BitMap& mask() const { return _mask; }
```
Comment (eme64) [inline src/hotspot/share/opto/phasetype.hpp:198]: You could additionally assert `is_valid` before returning the mask, just to ensure it is never called when invalid.

**Review by robcasloz — COMMENTED** [review]
Looks good, but I think it would be clearer to use the term `set` instead of `mask` through the entire changeset, to communicate the intent of the data structure rather than how it is implemented.

File: src/hotspot/share/opto/phasetype.hpp (hunk 3 above)
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp]: True, fixed now.

File: src/hotspot/share/compiler/directivesParser.cpp (hunk 4 above)
Comment (dlunde) [reply src/hotspot/share/compiler/directivesParser.cpp]: Do you mean extracting the mask from the validator and then passing it to `set_ideal_phase_mask`? Isn't it simpler to just pass the validator and then handle it internally in `set_ideal_phase_mask`?

I've also renamed the above after Roberto's suggestion, so it's now:
```
set->set_ideal_phase_name_set(validator);
```

File: src/hotspot/share/opto/phasetype.hpp (hunk 5 above)
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp]: Alright, sounds good to me (now changed). The previous style is the same as for `Compile::Compile` in `compile.cpp`, but the style differs between different parts of HotSpot.

File: src/hotspot/share/opto/phasetype.hpp (hunk 6 above)
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp]: Thanks, added now.

File: src/hotspot/share/compiler/compilerDirectives.cpp [hunk 7]
```diff
@@ -297,8 +297,7 @@ void DirectiveSet::init_control_intrinsic() {
   }
 }
 
-DirectiveSet::DirectiveSet(CompilerDirectives* d) :_inlinematchers(nullptr), _directive(d) {
-  _ideal_phase_name_mask = 0;
+DirectiveSet::DirectiveSet(CompilerDirectives* d) :_inlinematchers(nullptr), _directive(d), _ideal_phase_name_set(PHASE_NUM_TYPES, mtCompiler) {
```
Comment (eme64) [inline src/hotspot/share/compiler/compilerDirectives.cpp:300]: Needs a space after the colon `:`. But even better would be to put the initializations on a new line each, with 2 spaces indentation, and the opening bracket on a new line by itself.

File: src/hotspot/share/compiler/directivesParser.cpp (hunk 4 above)
Comment (eme64) [reply src/hotspot/share/compiler/directivesParser.cpp]: Yes, extract the mask/set first, and pass it directly. There is no need for `set_ideal_phase_name_set` to get more than it asks for as input.

File: src/hotspot/share/compiler/compilerDirectives.cpp [hunk 8]
```diff
@@ -621,7 +619,7 @@ DirectiveSet* DirectiveSet::clone(DirectiveSet const* src) {
 #undef copy_string_members_definition
 
   set->_intrinsic_control_words = src->_intrinsic_control_words;
-  set->_ideal_phase_name_mask = src->_ideal_phase_name_mask;
+  set->_ideal_phase_name_set.set_from(src->_ideal_phase_name_set);
```
Comment (eme64) [inline src/hotspot/share/compiler/compilerDirectives.cpp:622]: The advantage of passing the set/mask directly is that here you could also use your `set_ideal_phase_name_set` method.

**Review by eme64 — CHANGES_REQUESTED** [review]
Looks much better already, just a few more comments :)

File: src/hotspot/share/compiler/compilerDirectives.cpp (hunk 7 above)
Comment (chhagedorn) [reply src/hotspot/share/compiler/compilerDirectives.cpp:300]: Unfortunately, there does not seem to be a rule to follow in our style guide about how to format a wrapped initializer list - maybe that's something to think about to add at some point. The Google C++ style guide [suggests](https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists) the following (which I often adhere to but I always put each member on a new line for clarity if it's too much to put on one line):
```
Constructor initializer lists can be all on one line or with subsequent lines indented four spaces.
```
```

File: src/hotspot/share/compiler/directivesParser.cpp (hunk 4 above)
Comment (dlunde) [reply src/hotspot/share/compiler/directivesParser.cpp]: OK, fine with me. Now changed.

File: src/hotspot/share/compiler/compilerDirectives.cpp (hunk 8 above)
Comment (dlunde) [reply src/hotspot/share/compiler/compilerDirectives.cpp]: Thanks, updated.

File: src/hotspot/share/compiler/compilerDirectives.cpp [hunk 9]
```diff
@@ -434,11 +437,10 @@ DirectiveSet* DirectiveSet::compilecommand_compatibility_init(const methodHandle
       // Parse ccstr and create mask
```
Comment (robcasloz) [inline src/hotspot/share/compiler/compilerDirectives.cpp:437]: Could you update (or remove) this comment as part of the renaming?
Please also update the comment on line 430:
`    // Parse PrintIdealPhaseName and create an efficient lookup mask`

**Review by robcasloz — CHANGES_REQUESTED** [review]
Thanks for doing the renaming, Daniel. I tried this out by creating more than 64 compiler phases and running `-XX:CompileCommand=PrintIdealPhase,...` with a few random phase sets and it worked just fine. Please have a look at my (hopefully) final comments.

File: src/hotspot/share/compiler/compilerDirectives.cpp (hunk 9 above)
Comment (dlunde) [reply src/hotspot/share/compiler/compilerDirectives.cpp]: Thanks, both updated now

File: src/hotspot/share/opto/phasetype.hpp [hunk 10]
```diff
@@ -157,11 +156,17 @@ class PhaseNameIter {
 
 class PhaseNameValidator {
  private:
+  CHeapBitMap _phase_name_set;
   bool _valid;
   char* _bad;
 
+
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp]: Thanks, removed

**Review by robcasloz — APPROVED** [review]

File: src/hotspot/share/opto/phasetype.hpp [hunk 11]
```diff
@@ -186,13 +190,12 @@ class PhaseNameValidator {
...
-  }
-
-  const char* what() const {
-    return _bad;
+  const BitMap& phase_name_set() const {
+    assert(is_valid(), "Use of invalid phase name set");
+    return _phase_name_set;
   }
+  bool is_valid() const { return _valid; }
+  const char* what() const { return _bad; }
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phasetype.hpp:198]: I suggest to either keep the old format with the return on a new line or add a new line between `phase_name_set()` and `is_valid()` to better separate it.

**Review by chhagedorn — APPROVED** [review]
Otherwise, looks good!

File: src/hotspot/share/opto/phasetype.hpp (hunk 11 above)
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp]: OK, reverted to old format with return on new line and new line separating the functions.

**Review by chhagedorn — APPROVED** [review]

**Review by robcasloz — APPROVED** [review]

**Review by eme64 — APPROVED** [review]

**Comment by dlunde** [comment]
Thanks everyone for the reviews. Integrating now (but I require a sponsor).

**Comment by eme64** [comment]
Thanks for the work @dlunde ! /sponsor

**Pull request closed without merging**